Design/y_e_data_storage_modes.md

```diff
@@ -0,0 +1,151 @@
+
+- `Histogram` currently provides functions for obtaining variances as well as standard deviations, e.g., `countVariances()` and `countStandardDeviations()`.
+  If the internal `E` vector were converted to variances by external means (as it is done currently), these access methods would do the wrong thing.
+
+- `HistogramData` provides means for converting between standard deviations and variances, in particular types that do the conversion automatically.
+  These should be used by client code.
+
+- If summing spectra were done using `Histogram::operator+()`, propagating the uncertainties would involve a `sqrt` computation for every `+` operation.
+  In the example of `DiffractionFocussing2` this could yield a significant overhead, and it is not obvious how to solve this:
+
+  - The algorithm could do the summing itself, summing `Y` and variances separately.
+    This would cause no performance loss, and the code would be reasonably clean, e.g.,
+
+    ```cpp
+    for (size_t i=0; i<inputWS.getNumberHistograms(); ++i) {
+      counts += inputWS.counts(i);
+      variances += inputWS.countVariances(i);
+    }
+    ```
+
+  - We could provide optimized means for summing many spectra as part of the `HistogramData` module.
+    For example,
+
+    ```cpp
+    namespace HistogramData {
+    Histogram sumHistograms(const std::vector<Histogram> &histograms);
+    }
+    ```
+
+    This would be a clean solution without performance loss.
+    Unfortunately, there will be an implicit performance and ressource issue if the input is an `EventWorkspace`!
+    Obtaining a `Histogram` from the input workspace will create the `Y` and `E` data on the fly (this needs to be done in any solution).
+    However, to create the vector of histograms as input for `sumHistograms()` we need to **store all histograms at the same time**.
+    This will increase memory consumption and reduce cache locality of the computation.
+
+
+## Count and frequency data
+
+### Status
+
+- Data in `Workspace2D` be represent counts or frequencies.
+- Conversions between the two are done with the algorithms `ConvertToDistribution` and `ConvertFromDistribution`, or the helper function `WorkspaceHelpers::makeDistribution()`.
+- `MatrixWorkspace::isDistribution()` can be used for checking the state, and `MatrixWorkspace::setDistribution()` for setting the corresponding flag (note that this does not actually transform the data!).
+- There is a validator that verifies the state of a workspace, `RawCountValidator`, however is is used by relatively few algorithms.
+- A considerable number of algorithms use `MatrixWorkspace::isDistribution()` to adapt their behavior, in most cases this adaption is a simple conversion to the desired data storage type.
+- For a large number of algorithms there are no checks, and it is not documented whether or not the work only with counts, only with frequencies, or both.
+
```

Comment:
@OwenArnold Yes, Ithink the points of import are easy to deal with. The flag can be removed if we define that histograms always store counts. My only concern is the lossy conversion to frequencies for point-distribution data.

We might consider adding an extra field to `Histogram` that stores a helper value that aids in making the conversion lossless:

- We always store counts in `Histogram`
- Client code needs point-data with frequencies, sue `Histogram(Points{1,2,3}, Frequencies{4,5,6}).
- To convert to `Counts` for the internal store, we need a bin-width, which is not uniquely defined since we have `Points`.
- Instead of lossy conversion (Mantid is currently taking the mid-points between points, but when you convert back to points you won't get the same):
  - convert to bin-edges using mid-points (just like now)
  - store points and counts, and the leftmost bin edge (the latter is not accessible on the public interface)
  - `Histogram::frequencies()` can now convert the counts back to frequencies on a *lossless* way, since we have the extra required value (in contrast to now)